### PR TITLE
Add a section about iodata to the IO docs

### LINKED ISSUE #8940: Document IO lists and IO data
We should have a section in the IO module about them.

### PR DESCRIPTION
**Pull request by whatyouhide**
Closes #8940.

### CODE REVIEW

File: lib/elixir/lib/io.ex [hunk 1]
```diff
@@ -32,6 +29,109 @@ defmodule IO do
   was last accessed. The position of files can be changed using the
   `:file.position/2` function.
 
+  ## Iolists and iodata
```
Comment (fertapric) [inline lib/elixir/lib/io.ex]: What do you think about "IO Data/IO Lists" or "IOData/IOLists"? I've seen the latter [in the Erlang/OTP codebase](https://github.com/erlang/otp/blob/master/lib/kernel/src/raw_file_io_list.erl#L74) 🙂

File: lib/elixir/lib/io.ex [hunk 2]
```diff
@@ -32,6 +29,109 @@ defmodule IO do
...
+
+      def email(username, domain) do
+        username <> "@" <> domain
+      end
+
+  In this function, creating the email address will copy the `username` and `domain`
+  binaries. Now imagine you want to use the resulting email inside another binary:
+
+      def welcome_message(name, username, domain) do
+        "Welcome #{name}, your email is: #{email(username, domain)}"
```
Comment (fertapric) [inline lib/elixir/lib/io.ex]: This heredoc allows interpolation.

File: lib/elixir/lib/io.ex [hunk 3]
```diff
@@ -32,6 +29,109 @@ defmodule IO do
   was last accessed. The position of files can be changed using the
   `:file.position/2` function.
 
+  ## Iolists and iodata
+
+  Iolists and iodata are data structures that can be used as a sometimes
+  more efficient alternative to binaries.
```
Comment (josevalim) [inline lib/elixir/lib/io.ex]: "that can be used as a more efficient alternative to binaries under certain situations".

File: lib/elixir/lib/io.ex (hunk 1 above)
Comment (josevalim) [reply lib/elixir/lib/io.ex]: Now that I think about it, we don't expose iolists anywhere in Elixir except in the typespecs. Most places where we say iolists informally we actually mean iodata. So my suggestion is to call it "IO data" and focus on that. We can have one note on iolists saying that iolists are the same as iodata except they don't accept a binary in the top level.

Comment (whatyouhide) [reply lib/elixir/lib/io.ex]: Good call on dropping the iolists. We should go with "IO data" for the naming IMO.

**Comment by whatyouhide** [comment]
@fertapric @josevalim updated with your suggestions. :)

File: lib/elixir/lib/io.ex [hunk 4]
```diff
@@ -1,19 +1,16 @@
...
-  lists of characters and strings. In case another type is given,
-  functions will convert to string via the `String.Chars` protocol
-  (as shown in typespecs).
-
-  The functions starting with `bin` expect iodata as an argument,
-  i.e. binaries or lists of bytes and binaries.
+  The majority of the functions expect chardata. In case another type is given,
+  functions will convert those types to string via the `String.Chars` protocol
+  (as shown in typespecs). For more information on chardata, see the
+  "Iolists and iodata" section below.
```
Comment (fertapric) [inline lib/elixir/lib/io.ex]:
```suggestion
  "IO data" section below.
```

Comment (whatyouhide) [reply lib/elixir/lib/io.ex]: Good catch!

**Pull request merged**